pkg/aws/secret.go

```diff
@@ -19,50 +19,79 @@ import (
 	"fmt"
 
 	awsclient "github.com/gardener/gardener-extension-provider-aws/pkg/aws/client"
-	extensionscontroller "github.com/gardener/gardener/extensions/pkg/controller"
 
+	extensionscontroller "github.com/gardener/gardener/extensions/pkg/controller"
 	corev1 "k8s.io/api/core/v1"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
 // GetCredentialsFromSecretRef reads the secret given by the the secret reference and returns the read Credentials
 // object.
-func GetCredentialsFromSecretRef(ctx context.Context, client client.Client, secretRef corev1.SecretReference) (*Credentials, error) {
+func GetCredentialsFromSecretRef(ctx context.Context, client client.Client, secretRef corev1.SecretReference, useDNSKeys bool) (*Credentials, error) {
 	secret, err := extensionscontroller.GetSecretByReference(ctx, client, &secretRef)
 	if err != nil {
 		return nil, err
 	}
-	return ReadCredentialsSecret(secret)
+	return ReadCredentialsSecret(secret, useDNSKeys)
 }
 
 // ReadCredentialsSecret reads a secret containing credentials.
-func ReadCredentialsSecret(secret *corev1.Secret) (*Credentials, error) {
+func ReadCredentialsSecret(secret *corev1.Secret, useDNSKeys bool) (*Credentials, error) {
 	if secret.Data == nil {
 		return nil, fmt.Errorf("secret does not contain any data")
 	}
 
-	accessKeyID, ok := secret.Data[AccessKeyID]
+	accessKeyIDKey, secretAccessKeyKey, regionKey := AccessKeyID, SecretAccessKey, Region
+	if useDNSKeys {
+		accessKeyIDKey, secretAccessKeyKey, regionKey = DNSAccessKeyID, DNSSecretAccessKey, DNSRegion
+	}
```

Comment:
I could do it, but I would like to be sure that there is a use case for this. If gardener would never pass DNS secrets in the second vairant, then the added complexity is not worth it.

> I remember, there is a use case in gardener, where you specify a non-default domain in Shoot.spec.dns.domain but no custom dns provider (secret). In that case the shoots credentials are used. 

If I specify a non-default domain while having a `default-domain` secret I get an error during validation:
```
Error from server (BadRequest): error when creating "dev/shoot-i500152-gcp.yaml": shoot uses a default domain but does not match expected scheme: <shoot-name>.<project-name>.<default-domain>
```

And if I delete the `default-domain` secret, the shoot is accepted, but the reconciliation fails immediately with:
```
Could not initialize a new operation for Shoot reconciliation: unable to figure out which secret should be used for dns.
```

Without a `default-domain` secret, the only way I could reconcile such a shoot I found so far is to add a provider (with a secret) and set `primary` to `true`. So I don't think what you describe is actually possible. Could I be missing something?

Having in mind that:
* `external-dns-management` would so far only accept the "DNS format" of secrets. 
* gardener doesn't perform any mapping of secret keys (and I think it shouldn't, since it's provider-specific). 

The only reason to accept the other secret format would be to allow users who never used `external-dns-management` before to specify a secret in this format for a custom provider. I am sceptical that this justifies the increased complexity of supporting both secret variants in all providers (this is just one of them).

Reply:
Note: I double checked, `external-dns-management` indeed supports the "other format" via "alternative keys", e.g.:
```
accessKeyID, err := c.GetRequiredProperty("AWS_ACCESS_KEY_ID", "accessKeyID")
```
Still, the above consideration still applies. Is there a use case in which this format is actually used currently? If not, are we sure we want to introduce one?

Reply:
I double checked some more and found this code: https://github.com/gardener/gardener/blob/54930d430301a5f0aadfa5367e10012371c542d0/pkg/operation/shoot/shoot.go#L466-L475
Trying to test it now ...

Reply:
> I double checked some more and found this code: https://github.com/gardener/gardener/blob/54930d430301a5f0aadfa5367e10012371c542d0/pkg/operation/shoot/shoot.go#L466-L475
> Trying to test it now ...

Yep, exactly. Also had to dig a bit, but this is the use case that I remembered.
So, a non-default domain is used in conjunction with a primary dns provider that doesn't reference a secret.

Reply:
Done. Tested, works the use case mentioned above, works fine.